File: packages/cli/lib/files.ts
import { readFileSync, statSync, readdirSync, existsSync } from "fs";
import { join, basename, extname, relative, sep } from "path";
import { createHash } from "crypto";
import ignore from "ignore";

// Default patterns to ignore when scanning directories (fallback if no .gitignore)
const DEFAULT_IGNORE_PATTERNS = [
  ".git/",
  ".gitignore",
  "node_modules/",
  ".DS_Store",
  "Thumbs.db",
  ".env",
  ".env.*",
  "!.env.example",
  "*.log",
  ".cache/",
  "dist/",
  "build/",
  ".next/",
  ".vercel/",
  ".turbo/",
  "coverage/",
  ".nyc_output/",
];

export interface FileInfo {
  originalPath: string;
  path: string;
  size: number;
  sha: string;
  extension: string;
  content: Buffer;
  projectName?: string;
}

/**
 * Load and parse .gitignore file from a directory
 */
function loadGitignore(baseDir: string): ReturnType<typeof ignore> {
  const ig = ignore();

  // Add default patterns
  ig.add(DEFAULT_IGNORE_PATTERNS);

  // Look for .gitignore in the base directory
  const gitignorePath = join(baseDir, ".gitignore");

  if (existsSync(gitignorePath)) {
    try {
      const gitignoreContent = readFileSync(gitignorePath, "utf-8");
      ig.add(gitignoreContent);
    } catch (error) {
      console.warn(`Warning: Could not read .gitignore file: ${error}`);
    }
  }

  return ig;
}

/**
 * Check if a file path should be included
 */
function shouldIncludeFile(
  filePath: string,
  ig: ReturnType<typeof ignore>,
): boolean {
  // Normalize path separators for consistent matching
  const normalizedPath = filePath.split(sep).join("/");
  return !ig.ignores(normalizedPath);
}

/**
 * Calculate SHA hash of file content
 */
function calculateSha(content: Buffer): string {
  return createHash("sha1").update(content).digest("hex");
}

/**
 * Get file extension without the dot
 */
function getExtension(filePath: string): string {
  const ext = extname(filePath);
  return ext ? ext.slice(1).toLowerCase() : "";
}

/**
 * Normalize path to use forward slashes (for consistency across platforms)
 */
function normalizePath(path: string): string {
  return path.split(sep).join("/");
}

/**
 * Recursively scan a directory and return all files
 */
function scanDirectory(dirPath: string, baseDir: string = dirPath): string[] {
  const files: string[] = [];
  const ig = loadGitignore(baseDir);

  function scan(currentPath: string): void {
    const entries = readdirSync(currentPath, { withFileTypes: true });

    for (const entry of entries) {
      const fullPath = join(currentPath, entry.name);
      const relativePath = relative(baseDir, fullPath);

      if (!shouldIncludeFile(relativePath, ig)) {
        continue;
      }

      if (entry.isDirectory()) {
        scan(fullPath);
      } else if (entry.isFile()) {
        files.push(fullPath);
      }
    }
  }

  scan(dirPath);
  return files;
}

/**
 * Discover files from a path or array of paths (files or directories)
 * Returns array of file objects with metadata
 */
export function discoverFiles(inputPaths: string | string[]): FileInfo[] {
  // Normalize to array
  const paths = Array.isArray(inputPaths) ? inputPaths : [inputPaths];

  if (paths.length === 0) {
    throw new Error("No paths provided");
  }

  const allFiles: FileInfo[] = [];
  let projectName: string | null = null;

  // Process each path
  for (let i = 0; i < paths.length; i++) {
    const inputPath = paths[i];
    if (!inputPath) continue;

    const stats = statSync(inputPath);
    const isFirstPath = i === 0;

    if (stats.isFile()) {
      // Single file - keep original filename
      const extension = getExtension(inputPath);
      const fileName = basename(inputPath, extname(inputPath));
      const content = readFileSync(inputPath);
      const targetPath = basename(inputPath);

      // Infer project name from first file
      if (isFirstPath) {
        projectName = fileName;
      }

      allFiles.push({
        originalPath: inputPath,
        path: normalizePath(targetPath),
        size: stats.size,
        sha: calculateSha(content),
        extension,
        content,
      });
    } else if (stats.isDirectory()) {
      // Directory
      const dirName = basename(inputPath);
      const dirFiles = scanDirectory(inputPath);

      // If this is the first path, use directory name as project name
      if (isFirstPath && !projectName) {
        projectName = dirName;
      }

      for (const filePath of dirFiles) {
        const content = readFileSync(filePath);
        const extension = getExtension(filePath);
        const relativePath = relative(inputPath, filePath);
        const fileStats = statSync(filePath);

        allFiles.push({
          originalPath: filePath,
          path: normalizePath(relativePath),
          size: fileStats.size,
          sha: calculateSha(content),
          extension,
          content,
        });
      }
    } else {
      throw new Error(
        `Invalid path: ${inputPath} is neither a file nor a directory`,
      );
    }
  }

  // Store project name in first file
  if (allFiles.length > 0 && projectName) {
    allFiles[0]!.projectName = projectName;
  }

  return allFiles;
}

/**
 * Get project name from discovered files
 */
export function getProjectName(files: FileInfo[]): string {
  if (files.length === 0) {
    throw new Error("No files discovered");
  }
  const projectName = files[0]?.projectName;
  if (!projectName) {
    throw new Error("Project name not found in files");
  }
  return projectName;
}

/**
 * Validate that files were discovered
 */
export function validateFiles(files: FileInfo[]): boolean {
  if (files.length === 0) {
    throw new Error("No files found to publish");
  }

  const hasMarkdownOrHtml = files.some((f) =>
    ["md", "mdx", "html"].includes(f.extension),
  );
  if (!hasMarkdownOrHtml) {
    console.warn(
      "Warning: No markdown or html files found. The site will be empty.",
    );
  }

  return true;
}

File: apps/cli/internal/files/files.go
package files

import (
	"bufio"
	"crypto/sha1"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

var defaultIgnorePatterns = []string{
	".git/",
	".gitignore",
	".flowershow",
	"node_modules/",
	".DS_Store",
	"Thumbs.db",
	".env",
	".env.*",
	"*.log",
	".cache/",
	"dist/",
	"build/",
	".next/",
	".vercel/",
	".turbo/",
	"coverage/",
	".nyc_output/",
}

// FileInfo holds metadata and content for a discovered file.
type FileInfo struct {
	OriginalPath string
	Path         string // normalized relative path (forward slashes)
	Size         int64
	SHA          string
	Extension    string
	Content      []byte
	ProjectName  string // only set on first file
}

// ignoreList is a compiled set of gitignore-style patterns.
type ignoreList struct {
	patterns []pattern
}

type pattern struct {
	raw     string
	negate  bool
	dirOnly bool
	// segments after splitting on '/'
	parts []string
}

func compilePatterns(lines []string) *ignoreList {
	il := &ignoreList{}
	for _, line := range lines {
		line = strings.TrimRight(line, "\r")
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		negate := false
		if strings.HasPrefix(line, "!") {
			negate = true
			line = line[1:]
		}
		dirOnly := strings.HasSuffix(line, "/")
		if dirOnly {
			line = strings.TrimSuffix(line, "/")
		}
		parts := strings.Split(strings.Trim(line, "/"), "/")
		il.patterns = append(il.patterns, pattern{
			raw:     line,
			negate:  negate,
			dirOnly: dirOnly,
			parts:   parts,
		})
	}
	return il
}

func (il *ignoreList) ignores(relPath string, isDir bool) bool {
	matched := false
	for _, p := range il.patterns {
		if p.dirOnly && !isDir {
			continue
		}
		if matchPattern(p, relPath) {
			matched = !p.negate
		}
	}
	return matched
}

func matchPattern(p pattern, relPath string) bool {
	// Normalize to forward slashes
	relPath = strings.ReplaceAll(relPath, string(os.PathSeparator), "/")
	parts := strings.Split(relPath, "/")

	// Single-segment pattern: match against any path component or basename
	if len(p.parts) == 1 {
		seg := p.parts[0]
		// Check against every path component
		for _, part := range parts {
			if globMatch(seg, part) {
				return true
			}
		}
		return false
	}

	// Multi-segment pattern: try to match from any starting position
	patLen := len(p.parts)
	for start := 0; start <= len(parts)-patLen; start++ {
		match := true
		for i, seg := range p.parts {
			if !globMatch(seg, parts[start+i]) {
				match = false
				break
			}
		}
		if match {
			return true
		}
	}
	return false
}

// globMatch matches a single path segment against a glob pattern.
func globMatch(pattern, name string) bool {
	// Simple glob: only '*' wildcard supported within a segment
	if pattern == "*" {
		return true
	}
	if !strings.Contains(pattern, "*") {
		return pattern == name
	}
	// Use filepath.Match for glob within a segment
	matched, _ := filepath.Match(pattern, name)
	return matched
}

func loadIgnore(baseDir string) *ignoreList {
	patterns := make([]string, len(defaultIgnorePatterns))
	copy(patterns, defaultIgnorePatterns)

	gitignorePath := filepath.Join(baseDir, ".gitignore")
	if f, err := os.Open(gitignorePath); err == nil {
		defer f.Close()
		scanner := bufio.NewScanner(f)
		for scanner.Scan() {
			patterns = append(patterns, scanner.Text())
		}
	}

	return compilePatterns(patterns)
}

func sha1Hex(content []byte) string {
	h := sha1.New()
	h.Write(content)
	return fmt.Sprintf("%x", h.Sum(nil))
}

func getExtension(path string) string {
	ext := filepath.Ext(path)
	if ext == "" {
		return ""
	}
	return strings.ToLower(ext[1:])
}

func normalizePath(path string) string {
	return strings.ReplaceAll(path, string(os.PathSeparator), "/")
}

func scanDirectory(dirPath, baseDir string, ig *ignoreList) ([]string, error) {
	var files []string

	err := filepath.Walk(dirPath, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if path == dirPath {
			return nil
		}

		relPath, err := filepath.Rel(baseDir, path)
		if err != nil {
			return err
		}

		if ig.ignores(relPath, info.IsDir()) {
			if info.IsDir() {
				return filepath.SkipDir
			}
			return nil
		}

		if !info.IsDir() {
			files = append(files, path)
		}
		return nil
	})

	return files, err
}

// DiscoverFiles discovers files from one or more paths (files or directories).
func DiscoverFiles(inputPaths []string) ([]FileInfo, error) {
	if len(inputPaths) == 0 {
		return nil, fmt.Errorf("no paths provided")
	}

	var allFiles []FileInfo
	var projectName string

	for i, inputPath := range inputPaths {
		info, err := os.Stat(inputPath)
		if err != nil {
			return nil, fmt.Errorf("cannot access path %s: %w", inputPath, err)
		}

		isFirst := i == 0

		if info.Mode().IsRegular() {
			content, err := os.ReadFile(inputPath)
			if err != nil {
				return nil, err
			}
			base := filepath.Base(inputPath)
			ext := getExtension(inputPath)
			if isFirst {
				projectName = strings.TrimSuffix(base, filepath.Ext(base))
			}
			allFiles = append(allFiles, FileInfo{
				OriginalPath: inputPath,
				Path:         normalizePath(base),
				Size:         info.Size(),
				SHA:          sha1Hex(content),
				Extension:    ext,
				Content:      content,
			})
		} else if info.IsDir() {
			if isFirst && projectName == "" {
				projectName = filepath.Base(inputPath)
			}
			ig := loadIgnore(inputPath)
			filePaths, err := scanDirectory(inputPath, inputPath, ig)
			if err != nil {
				return nil, err
			}
			for _, filePath := range filePaths {
				content, err := os.ReadFile(filePath)
				if err != nil {
					return nil, err
				}
				fi, err := os.Stat(filePath)
				if err != nil {
					return nil, err
				}
				relPath, _ := filepath.Rel(inputPath, filePath)
				allFiles = append(allFiles, FileInfo{
					OriginalPath: filePath,
					Path:         normalizePath(relPath),
					Size:         fi.Size(),
					SHA:          sha1Hex(content),
					Extension:    getExtension(filePath),
					Content:      content,
				})
			}
		} else {
			return nil, fmt.Errorf("invalid path: %s is neither a file nor a directory", inputPath)
		}
	}

	if len(allFiles) > 0 && projectName != "" {
		allFiles[0].ProjectName = projectName
	}

	return allFiles, nil
}

// GetProjectName returns the project name derived from the first file.
func GetProjectName(files []FileInfo) (string, error) {
	if len(files) == 0 {
		return "", fmt.Errorf("no files discovered")
	}
	if files[0].ProjectName == "" {
		return "", fmt.Errorf("project name not found in files")
	}
	return files[0].ProjectName, nil
}

// ValidateFiles checks that files were discovered and warns if no markdown/html found.
func ValidateFiles(files []FileInfo) error {
	if len(files) == 0 {
		return fmt.Errorf("no files found to publish")
	}
	hasMarkup := false
	for _, f := range files {
		if f.Extension == "md" || f.Extension == "mdx" || f.Extension == "html" {
			hasMarkup = true
			break
		}
	}
	if !hasMarkup {
		fmt.Fprintln(os.Stderr, "Warning: No markdown or html files found. The site will be empty.")
	}
	return nil
}
